Implement scrollbar-width CSS property in WebKit Legacy

lukewarlow opened the pull request:
#### e55f3a44ffb20e7d84e5415465ee54436d29559a
<pre>
Implement scrollbar-width CSS property in WebKit Legacy
<a href="https://bugs.webkit.org/show_bug.cgi?id=257056">https://bugs.webkit.org/show_bug.cgi?id=257056</a>

Reviewed by Tim Nguyen.

This change implements the missing code for the scrollbar-width CSS property none value to work inside of WebKit Legacy.

The thin value doesn&apos;t always have an effect in WebKit Legacy but this is spec compliant.

* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-007-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/cssom-view/client-props-root-expected.txt: Removed.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebKitLegacy/mac/WebView/WebDynamicScrollBarsView.mm:
(-[WebDynamicScrollBarsView updateScrollers]):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView _scrollbarWidthStyle]):
* Source/WebKitLegacy/mac/WebView/WebHTMLViewInternal.h:

Canonical link: <a href="https://commits.webkit.org/264633@main">https://commits.webkit.org/264633@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a4fa03d7f25d985a0f9f8b3ef9e0dddf0f5fff0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8195 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8487 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8704 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9859 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8261 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8203 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10475 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8398 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11138 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8340 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9406 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7432 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9981 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6706 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7498 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15056 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7829 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7627 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10987 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8100 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6583 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7391 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1973 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11598 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7841 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/WebKitLegacy/mac/WebView/WebDynamicScrollBarsView.mm:
I assumed changing the 15 on lines 390 and 401 to 11 based on the scrollbarWidthStyle being thin would result in thin scrollbars but this doesn't work. I assume the thin value isn't working because somewhere scrollbarThickness is being used without correctly passing down the scrollbar width value.

Based on my understanding of the spec not obeying the thin value should be compliant. Happy to change the commit message and make a follow up issue to look into supporting thin inside of WebKit Legacy if that's not the case.

lukewarlow commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -6288,7 +6288,7 @@ OverscrollBehavior LocalFrameView::verticalOverscrollBehavior()  const
 {
     auto* document = m_frame->document();
-    auto scrollingObject = document && document->scrollingElement() ? document->scrollingElement()->renderer() : nullptr;
+    auto scrollingObject = document && document->documentElement() ? document->documentElement()->renderer() : nullptr;
```
A new WPT test has been added which tests to make sure the body style doesn't propagate (like scrollbar-width-008.html) but runs in quirks mode. That passes inside of Chrome (with appropriate flags), and Firefox but not WebKit (before this change).

When in quirks mode scrollingElement resolves to body, whereas in standards mode it resolves to documentElement. 

I originally had this using documentElement but it was changed during the review process for WebKit 2 implementation. Based on the interop issue caught by the new test I've made I believe I was originally correct and this should be documentElement.

nt1m replied:
Can you split this change out (along with the test) to a separate commit? Just to make it clear that it has nothing to do with making `scrollbar-width` work in WebKit Legacy.

lukewarlow replied:
moved to https://github.com/WebKit/WebKit/pull/14429

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-009.html:
```diff
@@ -0,0 +1,65 @@
+  <div id="expander"></div>
+
+  <script type="text/javascript">
+    setup({ explicit_done: true });
```
Not sure the setup() & done() calls are any useful here, I don't see anything async in this test.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-009.html:
```diff
@@ -0,0 +1,65 @@
+
+  <div id="expander"></div>
+
+  <script type="text/javascript">
```
nit: 
```suggestion
  <script>
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-009.html:
```diff
@@ -0,0 +1,65 @@
+<meta charset="utf-8">
+<title>CSS Scrollbars: scrollbar-width on the body is not propagated quirks mode</title>
+<link rel="author" title="Felipe Erias Morandeira" href="mailto:felipeerias@gmail.com" />
```
Is this the correct author?

lukewarlow replied:
My bad I forgot to change that when copying scrollbar-width-008.html I will address as part of the seperate PR

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-009.html:
```diff
@@ -0,0 +1,65 @@
+<meta charset="utf-8">
+<title>CSS Scrollbars: scrollbar-width on the body is not propagated quirks mode</title>
+<link rel="author" title="Felipe Erias Morandeira" href="mailto:felipeerias@gmail.com" />
+<link rel="help" href="https://www.w3.org/TR/css-scrollbars-1/" />
```
nit: please link to https://drafts.csswg.org/css-scrollbars-1/

lukewarlow commented on Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
```diff
@@ -6255,6 +6255,16 @@ - (WebPluginController *)_pluginController
+    if (!frame || !frame->document() || !frame->document()->documentElement() || !frame->document()->documentElement()->renderer())
+        return WebCore::ScrollbarWidth::Auto;
+
+    return frame->document()->documentElement()->renderer()->style().scrollbarWidth();
```
uses documentElement not scrollingElement as scrollingElement was wrong. Existing code fixed in https://github.com/WebKit/WebKit/pull/14429

lukewarlow committed:
Implement scrollbar-width CSS property in WebKit Legacy
https://bugs.webkit.org/show_bug.cgi?id=257056

Reviewed by Tim Nguyen.

This change implements the missing code for the scrollbar-width CSS property none value to work inside of WebKit Legacy.

The thin value doesn't always have an effect in WebKit Legacy but this is spec compliant.

* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-007-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/cssom-view/client-props-root-expected.txt: Removed.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebKitLegacy/mac/WebView/WebDynamicScrollBarsView.mm:
(-[WebDynamicScrollBarsView updateScrollers]):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView _scrollbarWidthStyle]):
* Source/WebKitLegacy/mac/WebView/WebHTMLViewInternal.h:

Canonical link: https://commits.webkit.org/264633@main
